v0/destinations/ga4/utils.js

```diff
@@ -0,0 +1,259 @@
+ * @param {*} userProperties
+ */
+function removeReservedUserPropertyPrefixNames(userProperties) {
+  const reservedPrefixesNames = ["google_", "ga_", "firebase_"];
+
+  if (!userProperties) {
+    return;
+  }
+
+  Object.keys(userProperties).forEach(key => {
+    const valFound = reservedPrefixesNames.some(prefix => {
+      if (key.toLowerCase().startsWith(prefix)) {
+        return true;
+      }
+      return false;
+    });
+
+    // reject if found
+    if (valFound) {
+      // eslint-disable-next-line no-param-reassign
+      delete userProperties[key];
+    }
+  });
+}
+
+/* For custom events */
+/**
+ * Reserved custom event names cannot be used (Web)
+ * Ref - https://support.google.com/analytics/answer/10085872#zippy=%2Creserved-prefixes-and-event-names%2Cweb
+ * @param {*} event
+ * @returns
+ */
+function isReservedWebCustomEventName(event) {
+  const reservedEventNames = [
+    "app_remove",
+    "app_store_refund",
+    "app_store_subscription_cancel",
+    "app_store_subscription_convert",
+    "app_store_subscription_renew",
+    "first_open",
+    "first_visit",
+    "in_app_purchase",
+    "session_start",
+    "user_engagement"
+  ];
+
+  return reservedEventNames.includes(event.toLowerCase());
+}
+
+/**
+ * Reserved custom event name cannot start with reserved prefixes (Web)
+ * Ref - https://support.google.com/analytics/answer/10085872#zippy=%2Creserved-prefixes-and-event-names%2Cweb
+ * @param {*} event
+ * @returns
+ */
+function isReservedWebCustomPrefixName(event) {
+  const reservedPrefixesNames = ["_", "firebase_", "ga_", "google_", "gtag."];
+
+  // As soon as a single true is returned, .some() will itself return true and stop
+  return reservedPrefixesNames.some(prefix => {
+    if (event.toLowerCase().startsWith(prefix)) {
+      return true;
+    }
+    return false;
+  });
+}
+
+const GA4_ITEM_EXCLUSION = [
+  "item_id",
+  "itemId",
+  "product_id",
+
+  "item_name",
+  "itemName",
+  "name",
+
+  "coupon",
+
+  "item_category",
+  "itemCategory",
+  "category",
+
+  "item_brand",
+  "itemBrand",
+  "brand",
+
+  "item_variant",
+  "itemVariant",
+  "variant",
+
+  "price",
+  "quantity",
+
+  "index",
+  "position"
+];
+
+/**
+ * get items properties for a event.
+ * @param {*} message
+ * @returns
+ */
+function getDestinationItemProperties(message, isItemsRequired) {
+  let items;
+  const products = get(message, "properties.products");
+  if ((!products && isItemsRequired) || products.length === 0) {
+    throw new CustomError(
+      `Products is an required field for '${message.event}' event`,
+      400
+    );
+  }
+  if (products && Array.isArray(products)) {
+    items = [];
+    products.forEach((item, index) => {
+      let element = constructPayload(
+        item,
+        mappingConfig[ConfigCategory.ITEMS.name]
+      );
+      if (!element.item_id && !element.item_name) {
+        throw new CustomError("One of product_id or name is required", 400);
+      }
+
```

Comment:
Don't discard the entire event for a single product.

Reply:
It is a required field. [Ref](https://developers.google.com/analytics/devguides/collection/protocol/ga4/reference/events#begin_checkout_item)

Reply:
Yes, that I understood. My concern is we are discarding the entire call for a single product being missed the required field. Can't we implement in that way, filer out the wrong products and send the correct products only?

Reply:
If we do this then the filtration will be done silently without the user being aware of the properties they are trying to send. IMO keeping this gives the user a better idea of what exactly needs to be passed to get proper reporting on the GA4 dashboard. cc @utsabc

Reply:
I agree with @akashrpo on this but !element.item_id will not work when item id is `0` we should use defined and not null

Reply:
Added this check